core/playlistmanager.go

```diff
@@ -0,0 +1,262 @@
+	"github.com/golang/glog"
+	"github.com/livepeer/go-livepeer/drivers"
+	"github.com/livepeer/go-livepeer/net"
+	ffmpeg "github.com/livepeer/lpms/ffmpeg"
+	"github.com/livepeer/lpms/stream"
+)
+
+const LIVE_LIST_LENGTH uint = 10
+
+// XXX what if list will be longer?
+const FULL_LIST_LENGTH uint = 131072
+
+/*
+	PlaylistManager manages playlists and data for one video stream, backed by one object storage.
+*/
+type PlaylistManager interface {
+	ManifestID() ManifestID
+	// Implicitly creates master and media playlists
+	InsertHLSSegment(streamID StreamID, seg *stream.HLSSegment) (*net.TypedURI, error)
+	// InsertLink link to segment that already in the storage
+	// (Transcoder wrote segment in B's owned storage and returned link, B adds this link to video source)
+	// returns true if it was acutally inserted
+	InsertLink(turi *net.TypedURI, duration float64) (bool, error)
+
+	GetHLSMasterPlaylist() *m3u8.MasterPlaylist
+
+	GetHLSMediaPlaylist(streamID StreamID) *m3u8.MediaPlaylist
+
+	GetSession() drivers.OSSession
+	// Ends object storage's session
+	EndSession()
+}
+
+type BasicPlaylistManager struct {
+	storageSession drivers.OSSession
+	manifestID     ManifestID
+	// Live playlist used for broadcasting
+	masterLivePList *m3u8.MasterPlaylist
+	// Full playlist is stored into OS for later reference
+	masterFullPList *m3u8.MasterPlaylist
+	mediaLiveLists  map[StreamID]*m3u8.MediaPlaylist
+	mediaFullLists  map[StreamID]*m3u8.MediaPlaylist
+	mapSync         *sync.Mutex
+}
+
+// NewBasicPlaylistManager create new BasicPlaylistManager struct
+func NewBasicPlaylistManager(manifestID ManifestID,
+	storageSession drivers.OSSession) *BasicPlaylistManager {
+
+	bplm := &BasicPlaylistManager{
+		storageSession:  storageSession,
+		manifestID:      manifestID,
+		masterLivePList: m3u8.NewMasterPlaylist(),
+		masterFullPList: m3u8.NewMasterPlaylist(),
+		mediaLiveLists:  make(map[StreamID]*m3u8.MediaPlaylist),
+		mediaFullLists:  make(map[StreamID]*m3u8.MediaPlaylist),
+		mapSync:         &sync.Mutex{},
+	}
+	return bplm
+}
+
+func (mgr *BasicPlaylistManager) ManifestID() ManifestID {
+	return mgr.manifestID
+}
+
+func (mgr *BasicPlaylistManager) GetSession() drivers.OSSession {
+	return mgr.storageSession
+}
+
+// InsertHLSSegment ...
+func (mgr *BasicPlaylistManager) InsertHLSSegment(streamID StreamID,
+	seg *stream.HLSSegment) (*net.TypedURI, error) {
+
+	if err := mgr.isRightStream(streamID); err != nil {
+		return nil, err
+	}
+	sid := string(streamID)
+	fmpl := mgr.getFullPL(streamID)
+	if fmpl == nil {
+		panic("no full media pl")
+	}
+	if mgr.hasSegment(fmpl, seg.SeqNo) {
+		return nil, fmt.Errorf("Already saved")
+	}
+	turi, _, err := mgr.storageSession.SaveData(sid, seg.Name, seg.Data)
+	if err != nil {
+		glog.Error("Error saving segment", err)
+	} else {
+		turi.SeqNo = seg.SeqNo
+		mgr.insertLink(turi, seg.Duration)
+	}
+	return turi, err
+}
+
+func (mgr *BasicPlaylistManager) isRightStream(streamID StreamID) error {
```

Comment:
For my own info, when would a different stream be potentially selected, necessitating us having to make this check? Is it because a single transcoder might be processing multiple jobs at a time, and therefore, could potentially be referring to another job?

Reply:
Yes, this one can be removed